src/Compilers/CSharp/Test/Semantic/Semantics/UnsafeTests.cs

```diff
@@ -333,9 +352,1053 @@ System.Collections.Generic.IEnumerator<int> Goo()
+        }
+
+        [Fact]
+        public void Iterator_UnsafeMethod_04()
+        {
+            var code = """
+                class C
+                {
+                    public unsafe System.Collections.Generic.IEnumerable<int> M()
+                    {
+                        int x = sizeof(S);
+                        System.Console.Write("I" + x);
+                        yield return x;
+                    }
+                }
+                #pragma warning disable CS0169 // The field 'S.f' is never used
+                struct S { int f; }
+                """;
+
+            CreateCompilation(code, parseOptions: TestOptions.Regular12, options: TestOptions.UnsafeReleaseDll).VerifyDiagnostics(
+                // (3,63): error CS8652: The feature 'ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public unsafe System.Collections.Generic.IEnumerable<int> M()
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "M").WithArguments("ref and unsafe in async and iterator methods").WithLocation(3, 63),
+                // (5,17): error CS8652: The feature 'ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         int x = sizeof(S);
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "sizeof(S)").WithArguments("ref and unsafe in async and iterator methods").WithLocation(5, 17));
+
+            var expectedDiagnostics = new[]
+            {
+                // (7,9): error CS9231: Cannot use 'yield return' in an 'unsafe' block
+                //         yield return x;
+                Diagnostic(ErrorCode.ERR_BadYieldInUnsafe, "yield").WithLocation(7, 9)
+            };
+
+            CreateCompilation(code, parseOptions: TestOptions.RegularNext, options: TestOptions.UnsafeReleaseDll).VerifyDiagnostics(expectedDiagnostics);
+            CreateCompilation(code, options: TestOptions.UnsafeReleaseDll).VerifyDiagnostics(expectedDiagnostics);
+        }
+
+        [Fact]
+        public void UnsafeContext_01()
+        {
+            var code = """
+                using System.Collections.Generic;
+                using System.Threading.Tasks;
+                unsafe class C
+                {
+                    IEnumerable<int> M()
+                    {
+                        yield return 1;
+                        var lam = async () => await Task.Yield();
+                        async void local() { await Task.Yield(); }
+                        local();
+                    }
+                }
+                """;
+
+            CreateCompilation(code, options: TestOptions.UnsafeReleaseDll, parseOptions: TestOptions.Regular12).VerifyDiagnostics(
+                // (8,31): error CS4004: Cannot await in an unsafe context
```

Comment:
>Cannot await in an unsafe context [](http://example.com/codeflow?start=41&length=33)

It seems somewhat surprising to me that an iterator within an unsafe context is a safe context, but, given that, it seems even more surprising that a nested function within the iterator is an unsafe context. That said, this looks like existing behavior.

Reply:
Yes, that's an existing behavior, probably a bug. I chose to fix it only in C#13+, that's also why the check exists as you asked in [https://github.com/dotnet/roslyn/pull/73046#discussion_r1569630788](https://github.com/dotnet/roslyn/pull/73046#discussion_r1569630788) - I can remove that langversion gate, but I'm unsure if that would be better.

Reply:
> Yes, that's an existing behavior, probably a bug. I chose to fix it only in C#13+

The spec for this feature should probably explicitly mention the C#12 safe/unsafe context for nested functions inside iterators, and the proposed behavior for C#13, if that's not already covered.